service/worker/scanner/executions/common/interfaces.go

```diff
@@ -0,0 +1,70 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package common
+
+type (
+	// InvariantManager represents a manager of several invariants.
+	// It can be used to run a group of invariant checks or fixes.
+	// It is responsible for running invariants in their dependency order.
+	InvariantManager interface {
+		RunChecks(Execution) CheckResult
+		RunFixes(Execution) FixResult
+		InvariantTypes() []InvariantType
+	}
+
+	// Invariant represents an invariant of a single execution.
+	// It can be used to check that the execution satisfies the invariant.
+	// It can also be used to fix the invariant for an execution.
+	Invariant interface {
+		Check(Execution, *InvariantResourceBag) CheckResult
+		Fix(Execution, *InvariantResourceBag) FixResult
```

Comment:
I'm worried that passing baggage between invariants is going to imply dependency and execution order, which will affect our ability to reliably turn on/off these invariants dynamically. I'd leave this out for now and bring it back when/if we decide to have dependencies between invariants. Even if we implement dependencies between invariants, I'd question the value of having this bag given that we only scan (and fix) up to one corruption per execution.

Reply:
Spoke about this offline.